### PR TITLE
PP-848: Ability to include requested custom resources in mom hook input file.

### PR DESCRIPTION
**Pull request by ashwathraop**
#### Bug/feature Description
* *Currently hook writers only get a limited amount of information in the hook input file. For example, if I request ngpus, I need to parse the exec_vnode attribute and then identify the node to find out if a gpu was requested for the node. If I want the software attribute or custom resources, I have to query the server for it. This can have a negative impact on the server and introduces potential hangs. I would like for all the information requested by the job or set by the server using defaults to be available in the hook input file.*
* *Issue ID: [PP-848](https://pbspro.atlassian.net/browse/PP-848)*

#### Affected Platform(s)
* *All*

#### Cause / Analysis / Design
* *PBS does not send custom resources associated with jobs to MoMs.*

#### Solution Description
* *Introduce a new flag "m" for custom resources which will allow PBS to send those custom resources defined with the aforementioned flag to MoMs*
* EDD can be found here: [EDD](https://pbspro.atlassian.net/wiki/spaces/PD/pages/1034158087/Ability+to+include+requested+custom+resources+in+mom+hook+input+file.)

#### Testing logs/output
[ptl_log.txt](https://github.com/PBSPro/pbspro/files/3020083/ptl_log.txt)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/pbs_smoketest.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ class SmokeTest(PBSTestSuite):
     # Class variables
     resc_types = [None, 'long', 'float', 'boolean', 'size', 'string',
                   'string_array']
-    resc_flags = [None, 'n', 'h', 'nh', 'q', 'f', 'fh']
+    resc_flags = [None, 'n', 'h', 'nh', 'q', 'f', 'fh', 'm']
```
Comment (mike0042) [inline test/tests/pbs_smoketest.py]: Might "mh" also be a valid combination?

**Comment by subhasisb** [comment]
@ashwathraop we should think of extending this to all attributes. And if we do that then we probably may not need this "send_job" vehicle? I think we need to consider this along with the overall flow here.

File: test/tests/pbs_smoketest.py (hunk 1 above)
Comment (ashwathraop) [reply test/tests/pbs_smoketest.py]: Yes. Added mh to the list now.

**Comment by ashwathraop** [comment]
@subhasisb I would recommend having separate PR for sending server/queue attributes to MoMs as we might need a different approach for that. For resources associated with jobs, it makes sense to go through send job route. What do you think?

**Comment by ashwathraop** [comment]
[ptl_smoke.txt](https://github.com/PBSPro/pbspro/files/3139211/ptl_smoke.txt)
[ptl.txt](https://github.com/PBSPro/pbspro/files/3139213/ptl.txt)
Test run after pep8 and adding 'mh' flag to smoke test.

**Comment by subhasisb** [comment]
> @subhasisb I would recommend having separate PR for sending server/queue attributes to MoMs as we might need a different approach for that. For resources associated with jobs, it makes sense to go through send job route. What do you think?

Agreed! For the job related attribute, sending them with send_job sounds good. For the other, non-job, attributes/resources, we will need a separate mechanism. However, two things to confirm before u go ahead:

1. To be consistent in our design, the "m" flag would make sense on ALL kinds of attributes/resources? (we would not want to introduce a flag called "m" for one kind of attributes and that is called something else, or has a separate syntax for other attributes. Hence the design for the entire thing must be proposed and reviewed, hopefully in one place.

2. We have to be certain that we are not doing to double send the jobs related attributes if we create "another" mechanism to send the rest of the attributes. Thus, a bit of thought about what the other mechanism is would be good, so that we are sure about this approach before we go ahead and commit.

Does this make sense?

**Comment by billnitzberg** [comment]
> Agreed! For the job related attribute, sending them with send_job sounds good. For the other, non-job, attributes/resources, we will need a separate mechanism. However, two things to confirm before u go ahead:

The 'm' flag makes sense for resources, and it's clear it fits with the current overall design of PBS Pro, in part, because there is already an existing interface to specify resource flags.   Anything beyond this should be in a separate PR.

Regarding 'm' for attributes, we need to be more careful, as there is no supported interface for changing attribute flags (outside the source code) today.  The design for that new interface (for modifying attribute flags outside of source code) should consider all the flags and all the types of attributes, not just the new 'm' one.   It makes sense to do this more carefully and then decide how much of that broader design should be implemented first, and all this should be a separate PR.

File: test/fw/ptl/utils/pbs_snaputils.py [hunk 2]
```diff
@@ -632,6 +632,7 @@ def __convert_flag_to_numeric(flag):
         ATR_DFLAG_OPWR = 0x08
         ATR_DFLAG_MGRD = 0x10
         ATR_DFLAG_MGWR = 0x20
+        ATR_DFLAG_MOM = 0x400
```
Comment (dilip-krishnan) [inline test/fw/ptl/utils/pbs_snaputils.py]: Just wondering why 0x400 and not 0x40,  it makes more sense to utilize bit in incremental order.

Comment (ashwathraop) [reply test/fw/ptl/utils/pbs_snaputils.py]: @dilip-krishnan existing definitions match with what we have in pbs_internal.h. For example we have ATR_DFLAG_RASSN = 0x4000. I tried to follow it and defined the same value for ATR_DFLAG_MOM as defined in pbs_internal.h.

Comment (mike0042) [reply test/fw/ptl/utils/pbs_snaputils.py]: @ashwathraop please add a comment to this effect. Also note that the value of ATR_DFLAG_RASSN is defined as 0x4000 and not 0x400 in pbs_internal.h.

Comment (ashwathraop) [reply test/fw/ptl/utils/pbs_snaputils.py]: pbs_internal.h has value of ATR_DFLAG_RASSN is 0x4000 and ATR_DFLAG_MOM is 0x400. That is how I defined. I am not sure if I follow your comment.

Here is the list of flags we used in the snaputils file:
        ATR_DFLAG_USRD = 0x01
        ATR_DFLAG_USWR = 0x02
        ATR_DFLAG_OPRD = 0x04
        ATR_DFLAG_OPWR = 0x08
        ATR_DFLAG_MGRD = 0x10
        ATR_DFLAG_MGWR = 0x20
        ATR_DFLAG_MOM = 0x400
        ATR_DFLAG_RASSN = 0x4000
        ATR_DFLAG_ANASSN = 0x8000
        ATR_DFLAG_FNASSN = 0x10000
        ATR_DFLAG_CVTSLT = 0x20000

Comment (mike0042) [reply test/fw/ptl/utils/pbs_snaputils.py]: Sorry @ashwathraop, I was looking at the wrong line. Your definition of ATR_DFLAG_MOM is correct. Please do add a comment indicating where the value originated.

**Comment by ashwathraop** [comment]
Added new ptl test case for combination of i and m flags.
[ptl_logs.txt](https://github.com/PBSPro/pbspro/files/3272980/ptl_logs.txt)

File: test/fw/ptl/utils/pbs_snaputils.py (hunk 2 above)
Comment (ashwathraop) [reply test/fw/ptl/utils/pbs_snaputils.py]: @mike0042 I have added the comment to refer pbs_internal.h

**Comment by bhroam** [comment]
@ashwathraop why did you disallow m and i?  The i flag means invisible to the user.  Hooks run as managers.  They should see the i flag resources.

**Comment by ashwathraop** [comment]
@bhroam I thought with 'i' flag no external interface can see the resource. I was wrong. Even today mom hooks can access the resources flagged with 'i'. I tried a simple example hook and I could see the information being updated on the job.
----------------------------------------------------
create resource test
set resource test type = string
set resource test flag = i
03:06:04[root: /home/ashr/work]# cat end.py 
import pbs
this_event = pbs.event()
this_job = this_event.job
this_job.resources_used["test"] = "hellojob"
03:06:11[root: /home/ashr/work]# qstat -xf 1002|grep test
    resources_used.test = hellojob
----------------------------------------------------
It is just that any update made to the custom resources at server level or requested at job submission won't reach mom but the changes done at the hooks can be seen by the managers. I guess I do not see any harm in allowing flag 'm' to go along with 'i'. @subhasisb what do you think?

**Comment by subhasisb** [comment]
> @bhroam I thought with 'i' flag no external interface can see the resource. I was wrong. Even today mom hooks can access the resources flagged with 'i'. I tried a simple example hook and I could see the information being updated on the job.
> Example:
> create resource test
> set resource test type = string
> set resource test flag = i
> 
> 03:06:04[root: /home/ashr/work]# cat end.py
> import pbs
> this_event = pbs.event()
> this_job = this_event.job
> this_job.resources_used["test"] = "hellojob"
> 
> 03:06:11[root: /home/ashr/work]# qstat -xf 1002|grep test
> resources_used.test = hellojob
> 
> Without 'm' flag today it is just that any update made to the custom resources at server level or requested at job submission won't reach mom but the changes done at the hooks can be seen by the managers. I guess I do not see any harm in allowing flag 'm' to go along with 'i'. @subhasisb what do you think?

After our offline discussion, i agree with Bhroam and you. And i see you have made the changes, so i am good with the PR.

**Comment by ashwathraop** [comment]
[ptl_smoke.txt](https://github.com/PBSPro/pbspro/files/3279609/ptl_smoke.txt)
Results after reverting the commit.

**Comment by ashwathraop** [comment]
@bhroam I reverted the changes. Can you look at it now?

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class Test_Custom_Resource_Perm(TestFunctional):
+    rsc_list = ['foo_str', 'foo_strm', 'foo_strh']
+    hook_list = ["start", "begin"]
+
+    def tearDown(self):
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: Everything you are reverting in tearDown() will be reverted by PTL.  I don't see the purpose of doing it here.

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            self.server.manager(MGR_CMD_DELETE, HOOK, id=hook, runas=ROOT_USER)
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        for rsc in self.rsc_list:
+            try:
+                self.server.manager(MGR_CMD_DELETE, RSC,
+                                    id=rsc, runas=ROOT_USER)
+            except PbsManagerError:
+                pass
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: Why are you deleting resources?  The system will be in the default state after TestFunctional.setUp() runs.

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+        self.momA = self.moms.values()[0]
+        self.momA.restart()
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: Why are you restarting the mom?

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        self.momA.delete_vnode_defs()
+        self.hostA = self.momA.shortname
+
+        rc = self.server.manager(MGR_CMD_DELETE, NODE, None, "")
+        self.assertEqual(rc, 0)
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: manager() wil raise an exception if it fails.  There is no need to check its return code. (and other places)

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        rc = self.server.manager(MGR_CMD_CREATE, NODE, id=self.hostA)
+        self.assertEqual(rc, 0)
+        self.server.expect(NODE, {'state': 'free'}, id=self.hostA)
+
+        # Next set custom resources with m flag and without m flag
+        # via qmgr -c 'create resource'
+        attr = {'type': 'string', 'flag': 'h'}
+        r = 'foo_str'
+        rc = self.server.manager(
+            MGR_CMD_CREATE, RSC, attr, id=r, runas=ROOT_USER, logerr=False)
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: The person running PTL is a manager, you don't need to run as root.

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        hook_name = "start"
+        a = {'event': "exechost_startup", 'enabled': 'True'}
+        rv = self.server.create_import_hook(
+            hook_name,
+            a,
+            startup_hook_body,
+            overwrite=True)
+        self.assertTrue(rv)
+
+        self.momA.restart()
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: Why are you restarting the mom?  The hook will be sent to the mom after it is created.

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        using a execjob_begin hook.
+        """
+        self.logger.info("test_custom_resc__single_node")
+
+        a = {'Resource_List.foo_str': 'str_noperm',
+             'Resource_List.foo_strm': 'str_perm',
+             'Resource_List.foo_strh': 'str_hook'
+             }
+        j = Job(TEST_USER)
+        j.set_attributes(a)
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: Consider making this more concise by setting the attributes in the Job() call.

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        j.set_sleep_time("100")
+        jid = self.server.submit(j)
+
+        self.server.expect(JOB, {
+            'job_state': 'R',
+            'Resource_List.foo_str': 'str_noperm',
+            'Resource_List.foo_strm': 'str_perm',
+            'Resource_List.foo_strh': 'str_hook'},
+            offset=1, attrop=PTL_AND, id=jid)
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: attrop=PTL_AND is the default now.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 8 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: Well, this restart is needed to run a startup hook on the MoM which creates a custom resource.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 7 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: I will remove it.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 6 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: Okay, makes sense. I will remove it.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 5 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: Not exactly needed. I will remove it.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 4 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: I will remove it.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 3 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: Okay. I will remove the teardown().

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 10 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: Okay. I will remove it.

**Comment by ashwathraop** [comment]
@bhroam Thank you for the comments on the ptl script. I have fixed them and here is the result from the latest run.
[ptl_log.txt](https://github.com/PBSPro/pbspro/files/3283179/ptl_log.txt)

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 9 above)
Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: Done.

File: test/tests/functional/pbs_resc_custom_perm.py (hunk 8 above)
Comment (bhroam) [reply test/tests/functional/pbs_resc_custom_perm.py]: Ahh, got it.  I didn't think about that.  Thanks for the explanation.

File: test/tests/functional/pbs_resc_custom_perm.py [hunk 11]
```diff
@@ -79,15 +60,13 @@ def setUp(self):
...
+        self.server.manager(
+            MGR_CMD_CREATE, RSC, attr, id=r, logerr=False)
 
         attr = {'type': 'string', 'flag': 'hm'}
         r = 'foo_strm'
-        rc = self.server.manager(
-            MGR_CMD_CREATE, RSC, attr, id=r, runas=ROOT_USER, logerr=False)
-        self.assertEqual(rc, 0)
+        self.server.manager(
+            MGR_CMD_CREATE, RSC, attr, id=r, logerr=False)
```
Comment (bhroam) [inline test/tests/functional/pbs_resc_custom_perm.py]: Not sure, but you might be able to merge these back to 1 line since they are shorter now.

Comment (ashwathraop) [reply test/tests/functional/pbs_resc_custom_perm.py]: It is done.

**Comment by ashwathraop** [comment]
I rebased and merged the commits as well. Once the checks are done PR can be merged.

**Pull request merged**